receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,319 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusexecreceiver
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	sdconfig "github.com/prometheus/prometheus/discovery/config"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/receiver/prometheusreceiver"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusexecreceiver/subprocessmanager"
+)
+
+const (
+	// template for port in strings
+	portTemplate string = "{{port}}"
+	// healthyProcessTime is the default time a process needs to stay alive to be considered healthy
+	healthyProcessTime time.Duration = 30 * time.Minute
+	// healthyCrashCount is the amount of times a process can crash (within the healthyProcessTime) before being considered unstable - it may be trying to find a port
+	healthyCrashCount int = 3
+	// delayMutiplier is the factor by which the delay scales
+	delayMultiplier float64 = 2.0
+	// initialDelay is the initial delay before a process is restarted
+	initialDelay time.Duration = 1 * time.Second
+)
+
+type prometheusExecReceiver struct {
+	params   component.ReceiverCreateParams
+	config   *Config
+	consumer consumer.MetricsConsumer
+
+	// Prometheus receiver config
+	promReceiverConfig *prometheusreceiver.Config
+
+	// Subprocess data
+	subprocessConfig *subprocessmanager.SubprocessConfig
+	port             int
+
+	// Underlying receiver data
+	prometheusReceiver component.MetricsReceiver
+
+	// Shutdown channel
+	shutdownCh chan struct{}
+}
+
+type runResult struct {
+	elapsed       time.Duration
+	subprocessErr error
+}
+
+// new returns a prometheusExecReceiver
+func new(params component.ReceiverCreateParams, config *Config, consumer consumer.MetricsConsumer) (*prometheusExecReceiver, error) {
+	if config.SubprocessConfig.Command == "" {
+		return nil, fmt.Errorf("no command to execute entered in config file for %v", config.Name())
+	}
+	subprocessConfig := getSubprocessConfig(config)
+	promReceiverConfig := getPromReceiverConfig(config)
+
+	return &prometheusExecReceiver{
+		params:             params,
+		config:             config,
+		consumer:           consumer,
+		subprocessConfig:   subprocessConfig,
+		promReceiverConfig: promReceiverConfig,
+		port:               config.Port,
+	}, nil
+}
+
+// Start creates the configs and calls the function that handles the prometheus_exec receiver
+func (per *prometheusExecReceiver) Start(ctx context.Context, host component.Host) error {
```

Comment:
Nit: I'd move this function below getSubprocessConfig, getPromReceiverConfig & extractName so related logic is closer together